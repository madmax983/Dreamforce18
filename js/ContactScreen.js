/*
 * Copyright (c) 2015-present, salesforce.com, inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided
 * that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this list of conditions and the
 * following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and
 * the following disclaimer in the documentation and/or other materials provided with the distribution.
 *
 * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to endorse or
 * promote products derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import {
    ScrollView,
    StyleSheet,
    View,
    TouchableHighlight,
    Text,
    Button
} from 'react-native';

import Field from './Field';
import storeMgr from './StoreMgr';

const onSave = navigation => {
    const contact = navigation.getParam("contact");
    contact.__locally_updated__ = contact.__local__ = true;
    storeMgr.saveContact(contact, () => navigation.pop());
}

class ContactScreen extends React.Component {
    static navigationOptions = navigation => {
        return {
            headerRight: (
                <View style={{flex: 0.5, flexDirection: "row"}}>
                    <Button title={"Save"} onPress={(e) => {onSave(navigation.navigation)}}/>
                </View>
            )
        }
    }

    constructor(props) {
        super(props);
        this.state = {
            contact: this.props.navigation.getParam('contact')
        }
    }

    onChange = (fieldKey, fieldValue) => {
        const contact = this.state.contact;
        contact[fieldKey] = fieldValue;
        this.setState({contact});
    }

    onDeleteUndeleteContact = navigation => {
        const contact = this.state.contact;
        contact.__locally_deleted__ = !contact.__locally_deleted__;
        contact.__local__ = contact.__locally_deleted__ || contact.__locally_updated__ || contact.__locally_created__;
        storeMgr.saveContact(contact, () => {
            navigation.pop()
        });
    }

    renderDeleteUndeleteButton = () => {
        let iconName = 'delete';
        let title = 'Delete';
        let bgColor = 'red';
        if (this.state.contact.__locally_deleted__) {
            iconName = 'delete-restore';
            title = 'Undelete';
            bgColor = 'blue';
        }

        return (
            <View style={{marginTop:10}}>
                <Button
                    backgroundColor={bgColor}
                    containerStyle={{alignItems:'stretch'}}
                    icon={{name: iconName, type: 'material-community'}}
                    title={title}
                    onPress={(e) => {this.onDeleteUndeleteContact(this.props.navigation)}}
                />
            </View>
        );
    }

    render() {
        return (
            <ScrollView>
                <View style={this.props.style}>
                    <Field fieldLabel="First name" fieldValue={this.state.contact.FirstName} onChange={(text) => this.onChange("FirstName", text)}/>
                    <Field fieldLabel="Last name" fieldValue={this.state.contact.LastName} onChange={(text) => this.onChange("LastName", text)}/>
                    <Field fieldLabel="Title" fieldValue={this.state.contact.Title} onChange={(text) => this.onChange("Title", text)}/>
                    <Field fieldLabel="Mobile phone" fieldValue={this.state.contact.MobilePhone} onChange={(text) => this.onChange("MobilePhone", text)}/>
                    <Field fieldLabel="Email address" fieldValue={this.state.contact.Email} onChange={(text) => this.onChange("Email", text)}/>
                    <Field fieldLabel="Department" fieldValue={this.state.contact.Department} onChange={(text) => this.onChange("Department", text)}/>
                    <Field fieldLabel="Home phone" fieldValue={this.state.contact.HomePhone} onChange={(text) => this.onChange("HomePhone", text)}/>
                    {this.renderDeleteUndeleteButton()}
                </View>
            </ScrollView>
        );
    }
}

export default ContactScreen;
