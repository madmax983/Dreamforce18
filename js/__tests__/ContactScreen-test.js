import React from 'react';
import ContactScreen from '../ContactScreen.js';

import renderer from 'react-test-renderer';

test('renders correctly', () => {
    const navigation = {
        getParam: function() {
            return( {
                FirstName: "Obi Wan",
                LastName: "Kenobi",
                Title: "Jedi Master",
                MobilePhone: "(999) 999-9999",
                Department: "The Force",
                Email: "obiwan@moseisley.net",
                HomePhone: "(111) 111-1111"
            });
        }
    };
    const tree = renderer.create(<ContactScreen navigation={navigation} />).toJSON();
    console.log(tree);
    expect(tree).toMatchSnapshot();
});