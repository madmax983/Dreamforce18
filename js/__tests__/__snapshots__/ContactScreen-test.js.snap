// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView>
  <View>
    <View
      style={undefined}
    >
      <View>
        <View
          style={
            Array [
              Object {},
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "#86939e",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "marginBottom": 1,
                    "marginLeft": 20,
                    "marginRight": 20,
                    "marginTop": 15,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            First name
          </Text>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "borderBottomColor": "#bdc6cf",
              "borderBottomWidth": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            style={
              Array [
                Object {
                  "color": "#86939e",
                  "minHeight": 36,
                  "width": 750,
                },
                Object {
                  "fontSize": 17.5,
                },
                undefined,
              ]
            }
            value="Obi Wan"
          />
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {},
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "#86939e",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "marginBottom": 1,
                    "marginLeft": 20,
                    "marginRight": 20,
                    "marginTop": 15,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Last name
          </Text>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "borderBottomColor": "#bdc6cf",
              "borderBottomWidth": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            style={
              Array [
                Object {
                  "color": "#86939e",
                  "minHeight": 36,
                  "width": 750,
                },
                Object {
                  "fontSize": 17.5,
                },
                undefined,
              ]
            }
            value="Kenobi"
          />
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {},
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "#86939e",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "marginBottom": 1,
                    "marginLeft": 20,
                    "marginRight": 20,
                    "marginTop": 15,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Title
          </Text>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "borderBottomColor": "#bdc6cf",
              "borderBottomWidth": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            style={
              Array [
                Object {
                  "color": "#86939e",
                  "minHeight": 36,
                  "width": 750,
                },
                Object {
                  "fontSize": 17.5,
                },
                undefined,
              ]
            }
            value="Jedi Master"
          />
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {},
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "#86939e",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "marginBottom": 1,
                    "marginLeft": 20,
                    "marginRight": 20,
                    "marginTop": 15,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Mobile phone
          </Text>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "borderBottomColor": "#bdc6cf",
              "borderBottomWidth": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            style={
              Array [
                Object {
                  "color": "#86939e",
                  "minHeight": 36,
                  "width": 750,
                },
                Object {
                  "fontSize": 17.5,
                },
                undefined,
              ]
            }
            value="(999) 999-9999"
          />
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {},
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "#86939e",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "marginBottom": 1,
                    "marginLeft": 20,
                    "marginRight": 20,
                    "marginTop": 15,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Email address
          </Text>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "borderBottomColor": "#bdc6cf",
              "borderBottomWidth": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            style={
              Array [
                Object {
                  "color": "#86939e",
                  "minHeight": 36,
                  "width": 750,
                },
                Object {
                  "fontSize": 17.5,
                },
                undefined,
              ]
            }
            value="obiwan@moseisley.net"
          />
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {},
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "#86939e",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "marginBottom": 1,
                    "marginLeft": 20,
                    "marginRight": 20,
                    "marginTop": 15,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Department
          </Text>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "borderBottomColor": "#bdc6cf",
              "borderBottomWidth": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            style={
              Array [
                Object {
                  "color": "#86939e",
                  "minHeight": 36,
                  "width": 750,
                },
                Object {
                  "fontSize": 17.5,
                },
                undefined,
              ]
            }
            value="The Force"
          />
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {},
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "#86939e",
                    "fontSize": 15,
                    "fontWeight": "bold",
                    "marginBottom": 1,
                    "marginLeft": 20,
                    "marginRight": 20,
                    "marginTop": 15,
                  },
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Home phone
          </Text>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "borderBottomColor": "#bdc6cf",
              "borderBottomWidth": 1,
              "marginLeft": 20,
              "marginRight": 20,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            style={
              Array [
                Object {
                  "color": "#86939e",
                  "minHeight": 36,
                  "width": 750,
                },
                Object {
                  "fontSize": 17.5,
                },
                undefined,
              ]
            }
            value="(111) 111-1111"
          />
        </View>
      </View>
      <View
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityLabel={undefined}
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={undefined}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "padding": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Delete
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
